TriblerGUI/widgets/tablecontentmodel.py

```diff
@@ -92,19 +107,54 @@ def data(self, index, role):
 
     def add_items(self, new_data_items):
         super(TriblerContentModel, self).add_items(new_data_items)
-        # Build reverse mapping from infohashes to rows
+        # Build reverse mapping from unique ids to rows
         items_len = len(self.data_items)
         new_items_len = len(new_data_items)
         for i, item in enumerate(new_data_items):
-            if "infohash" in item and item['infohash'] not in self.infohashes:
-                self.infohashes[item["infohash"]] = items_len - new_items_len + i
+            if "infohash" in item and item['infohash'] not in self.item_uid_map:
+                self.item_uid_map[item["infohash"]] = items_len - new_items_len + i
+            elif "public_key" in item and "id" in item and item['public_key']+str(item['id']) not in self.item_uid_map:
+                self.item_uid_map[item['public_key']+str(item['id'])] = items_len - new_items_len + i
+
+    def add_items_to_top(self, new_items):
+        """
+        Add new items to the top of the content list in the model.
+        :param new_items: List of new items to be added
+        :return: None
+        """
+        new_items_map = {}
+        insert_index = 0
+        unique_items = []
+        for item in new_items:
+            item_uid = self.get_item_uid(item)
+
+            if item_uid and item_uid not in self.item_uid_map:
+                new_items_map[item_uid] = insert_index
+                unique_items.append(item)
+                insert_index += 1
+
+        # If no new items are found, skip
+        if not unique_items:
+            return
+
+        # Else to make space for new unique items, update the position of the existing items
+        for item in self.data_items:
+            old_item_uid = self.get_item_uid(item)
+            if old_item_uid in self.item_uid_map:
+                new_items_map[old_item_uid] = insert_index + self.item_uid_map[old_item_uid]
+
+        # Update the table
+        self.beginInsertRows(QModelIndex(), 0, len(self.data_items) + len(unique_items)-1)
```

Comment:
Please note that I had to fix the rows insertion in one of my latest PRs.. You should only really insert the *new* rows, see https://doc.qt.io/qt-5/qabstractitemmodel.html#beginInsertRows

https://github.com/ichorid/tribler/blob/d6ba575efc767db142df63a1334b97f8e289355a/TriblerGUI/widgets/tablecontentmodel.py#L148

https://github.com/Tribler/tribler/pull/4374/commits/d6ba575efc767db142df63a1334b97f8e289355a

Reply:
Good point